### PR TITLE
Added coverage to workflow

### PR DESCRIPTION
**Pull request by JacksonElia**
Okay I finally got everything working and I think it's a massive improvement:
1. Separates unit tests and integration tests, this is good for two reasons: makes everything run faster as the unit tests and integration tests run in parallel, and lets us split up their coverage (I decided not to have the integration tests have a coverage report as I believe we should really just care about it for the unit tests)
2. Made it so the github action for integration tests will fail after 3 minutes for an individual test, or 5 minutes for the entire job
3. Added coverage report for just the airbrakes/ dir
4. Made the unit tests fail if coverage is under 80%
5. Added an html report to show coverage. Because we're not using something like codecov, if you want to see it you have to go to the github action's artifacts, then download the zip file, extract it, and open `index.html`. I don't feel like setting codecov up, or getting it to use github pages, so for now this is good enough.

~I still don't know why/how the servo has coverage.~
I think the servo has coverage because we just use the mock pin factory

image

### CODE REVIEW

**Comment by JacksonElia** [comment]
I couldnt figure out how to make it so the integration tests didn't count for coverage, but I did manage to exclude the tests/ from the coverage report as who cares

**Comment by harshil21** [comment]
Can we just switch to codecov now itself, cause I want a CI check on coverage actually. `pytest-cov` also has basically the same feature, which outputs coverage on console. Also the numbers are quite deceiving cause we definitely have zero tests for the servo.

**Comment by JacksonElia** [comment]
Yeah idk what the problem with the numbers are, that is weird. Codecov uses this though, its just like a nice UI I think. I can look into pytest-cov

File: .coveragerc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+branch = True
+
+# Source files to measure
+source = airbrakes
+
+# Patterns of files to omit from coverage
+omit =
+    */tests/*
+    */site-packages/*
+    setup.py
```
Comment (harshil21) [inline .coveragerc:12]: we don't have a setup.py
```suggestion
```

File: .coveragerc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    airbrakes/mock/*
+
+[paths]
+source =
+    src/
+    */site-packages/
```
Comment (harshil21) [inline .coveragerc:18]: does this work?

File: .coveragerc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    # Have to re-enable the standard pragma
+    pragma: no cover
+
+    # Don't complain about missing debug-only code
+    def __repr__
+    if self\.debug
+
+    # Don't complain if tests don't hit defensive assertion code
+    raise AssertionError
+    raise NotImplementedError
```
Comment (harshil21) [inline .coveragerc:38]: NotImplementedError is only in display.py, and AssertionError is not used?

File: .coveragerc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# Don't report files with 100% coverage
+skip_covered = False
+
+# Don't report files with no executable code
+skip_empty = True
+
+# Exclude certain lines from coverage measurement
+exclude_lines =
+    # Have to re-enable the standard pragma
+    pragma: no cover
```
Comment (harshil21) [inline .coveragerc:30]: we don't use this

File: .coveragerc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    # Don't complain about missing debug-only code
+    def __repr__
+    if self\.debug
+
+    # Don't complain if tests don't hit defensive assertion code
+    raise AssertionError
+    raise NotImplementedError
+
+    # Don't complain if non-runnable code isn't run
+    if 0:
```
Comment (harshil21) [inline .coveragerc:41]:
```suggestion
```

File: .coveragerc
Comment (harshil21) [inline .coveragerc:1]: is it possible to move this into `pyproject.toml`?

File: .github/workflows/integration_tests.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      - '**'  # Matches all branch names, for PRs
+  push:
+    branches:
+      - 'main'  # Run tests on main branch
+
+jobs:
+  integration-tests:
+    name: integration-tests
+    runs-on: ${{matrix.os}}
+    timeout-minutes: 5
```
Comment (harshil21) [inline .github/workflows/integration_tests.yml:15]: bruh I was just about to ask you to add this, but looks like you're already ahead of me! 😄

File: .github/workflows/integration_tests.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+name: Integration Tests
```
Comment (harshil21) [inline .github/workflows/integration_tests.yml:1]:
```suggestion
name: Integration Test
```
Singular for now?

File: .github/workflows/integration_tests.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  pull_request:
+    branches:
+      - '**'  # Matches all branch names, for PRs
+  push:
+    branches:
+      - 'main'  # Run tests on main branch
+
+jobs:
+  integration-tests:
+    name: integration-tests
```
Comment (harshil21) [inline .github/workflows/integration_tests.yml:13]:
```suggestion
    name: integration-test
```

File: .github/workflows/unit_tests.yml [hunk 9]
```diff
@@ -1,36 +1,45 @@
...
-          cache-dependency-path: '**/requirements*.txt'
-      - name: Install dependencies
-        run: |
-          python -W ignore -m pip install --upgrade pip
-          python -W ignore -m pip install .
-          python -W ignore -m pip install .[dev]          
+  jobs:
+    pytest:
+      name: pytest
+      runs-on: ${{matrix.os}}
```
Comment (harshil21) [inline .github/workflows/unit_tests.yml:14]: can you add timeout-minutes here too? Arguably more important here.

**Review by harshil21 — COMMENTED** [review]
Excellent work!

Can you add the instructions to view the coverage inside a README in `tests/`?

File: .coveragerc (hunk 1 above)
Comment (JacksonElia) [reply .coveragerc:12]: lol all of this was boiler plate leave me alone

File: .coveragerc (hunk 3 above)
Comment (JacksonElia) [reply .coveragerc:38]: boiler plate code, we might do it in the future idk

File: .coveragerc (hunk 5 above)
Comment (JacksonElia) [reply .coveragerc:41]: okay yeah tbh this one is just dumb

File: .coveragerc
Comment (JacksonElia) [reply .coveragerc:1]: idk, i found more docs for using this file

File: .github/workflows/integration_tests.yml (hunk 7 above)
Comment (JacksonElia) [reply .github/workflows/integration_tests.yml:1]: its gonna be multiple tho eventually, right?

File: .github/workflows/unit_tests.yml (hunk 9 above)
Comment (JacksonElia) [reply .github/workflows/unit_tests.yml:14]: I'll do it for the entire job

**Comment by JacksonElia** [comment]
I'll just add the instructions on how to run it to our readme where we talk about pytest. That being said, you might have to double check if it works cause I can't really run tests locally.

File: .github/workflows/integration_tests.yml (hunk 7 above)
Comment (harshil21) [reply .github/workflows/integration_tests.yml:1]: We run one integration test for 2 cases, does that count as singular or plural? 🤔

File: .github/workflows/integration_tests.yml (hunk 8 above)
Comment (JacksonElia) [reply .github/workflows/integration_tests.yml:13]: no

File: README.md [hunk 10]
```diff
@@ -87,12 +87,12 @@ flowchart TD
     Airbrakes --> Servo
 
 ```
-| Type    | Color    | Examples                                                                                                         |
-|---------|----------|------------------------------------------------------------------------------------------------------------------|
-| Entry point |  ![#ac6600](https://via.placeholder.com/15/ac6600/000000?text=+) `#ac6600`   | Main.py
-| Classes | ![#44007e](https://via.placeholder.com/15/44007e/000000?text=+) `#44007e`  | Airbrakes Context, State, Logger, IMU, IMU Data Processor, Servo, Coast, Standby, Freefall, Landed, Motor Burn |
-| Methods |  ![#164b6c](https://via.placeholder.com/15/164b6c/000000?text=+) `#164b6c`  | update(), log(), extend_airbrakes(), retract_airbrakes()                                                       |
-| Outputs |  ![#044728](https://via.placeholder.com/15/044728/000000?text=+) `#044728`   | IMU Data Packet, Processed Data Packet, Logged Data Packet    
+| Type        | Color                                                                     | Examples                                                                                                       |
```
Comment (harshil21) [inline README.md:90]: was there a visual change here? it looks the same to me

File: .coveragerc
Comment (harshil21) [reply .coveragerc:1]: https://coverage.readthedocs.io/en/latest/config.html#sample-file

here are the docs @JacksonElia

File: README.md (hunk 10 above)
Comment (JacksonElia) [reply README.md:90]: yeah, it was just pycharm doing better formatting for it lol

**Comment by harshil21** [comment]
the artifacts don't upload rn, otherwise LGTM.

**Comment by JacksonElia** [comment]
Okay I fixed it. Everything should be good

**Review by harshil21 — APPROVED** [review]
LGTM!

**Pull request merged**